packages/relic/lib/src/headers/parser/headers_parser.dart

```diff
@@ -0,0 +1,276 @@
+/// including single values, multiple values, URIs, dates, integers, and
+/// booleans.
+class HeadersParser {
+  final HttpHeaders headers;
+  final bool strict;
+  final void Function(String, List<String>) onHeaderFailedToParse;
+
+  HeadersParser({
+    required this.headers,
+    required this.strict,
+    required this.onHeaderFailedToParse,
+  });
+
+  /// Parses a single header value associated with [key] and returns it as type
+  /// [T], or `null` if no value is found.
+  ///
+  /// - If the header has multiple values, returns the first one.
+  /// - Throws an [InvalidHeaderException] if `strict` is `true` and the header
+  /// `key` is present but has an unexpected or invalid value.
+  T? parseSingleValue<T>(
+    String key, {
+    required T Function(String) onParse,
+  }) {
+    var singleValue = headers.value(key);
+
+    if (singleValue != null) {
+      try {
+        return onParse(singleValue);
+      } catch (e) {
+        if (!strict) {
+          onHeaderFailedToParse(key, [singleValue]);
+          return null;
+        }
+
+        _throwException(e, key: key);
+      }
+    }
+
+    var multiValues = headers[key];
+    if (multiValues == null) return null;
+
+    singleValue = multiValues.firstOrNull;
+
+    if (singleValue != null && multiValues.length == 1) {
+      try {
+        return onParse(multiValues.first);
+      } catch (e) {
+        if (!strict) {
+          onHeaderFailedToParse(key, multiValues);
+          return null;
+        }
+
+        _throwException(e, key: key);
+      }
+    }
+
+    if (strict) {
+      throw InvalidHeaderException(
+        "Multiple values are not allowed",
+        headerType: key,
+      );
+    }
+
+    if (singleValue == null) return null;
+
+    try {
+      return onParse(singleValue);
+    } catch (e) {
+      onHeaderFailedToParse(key, multiValues);
+      return null;
+    }
+  }
+
+  /// Parses multiple header values associated with [key] and returns them as
+  /// type [T], or `null` if no values are found.
+  ///
+  /// - Splits each value by commas and trims whitespace.
+  /// - Throws an [InvalidHeaderException] if `strict` is `true` and the header
+  /// `key` is present but contains only empty or invalid values.
+  T? parseMultipleValue<T>(
+    String key, {
+    required T Function(List<String>) onParse,
+  }) {
+    List<String>? multiValues = headers[key];
+    if (multiValues == null) return null;
+
+    try {
+      return onParse(multiValues);
+    } catch (e) {
+      if (!strict) {
+        onHeaderFailedToParse(key, multiValues);
+        return null;
+      }
+
+      _throwException(e, key: key);
+    }
+  }
+
+  /// Parses a URI from the header value associated with [key], or returns
+  /// `null` if no valid URI is found.
+  ///
+  /// - Throws an [InvalidHeaderException] if `strict` is `true` and the header
+  /// `key` is present but contains an invalid URI.
+  Uri? parseUri(String key) {
```

Comment:
Question(NIT - maybe we should not do anything): These methods don't seem to fit on this abstraction level. There isn't really a difference between this and `CacheControlHeade``parse` except that one is a dart supplied type? 

On this class method we should maybe only supply the `parseSingleValue` and `parseMultipleValue` methods. And then pass these type parsers in as values to the `parseMultipleValue` or `parseSingleValue` methods? (They can still be shared between different headers)

Now we have bunbled pulling one or more values our of the header and parsing them.